### PR TITLE
Replace Subscription and TopicCache RateLimiting with Lazy

### PR DESCRIPTION
**Pull request by danielmarbach**
By talking to @mauroservienti I understood that the previous rate-limiting code was brought into the code base under the assumptions the message driven pubsub compat mode will run out of support in v8. This turned out to be not true unfortunately, and now we have this code in place that:

- Does effectively not use proper license headers to attribute the original source
- Did assume we are not targeting NET 4.7.2 (that's why the sources were copied) which we do now
- Allocates non-trivial amount of closures
- Takes a single global semaphore with one slot per topic or subscription cache effectively serializing accesses to the cache for a brief period of time for all concurrent publishes. 

As a first attempt, I replaced the current rate-limiting code with [System.Threading.RateLimiting](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/) but it turned out we are not actually doing rate limiting, but we want to make sure only one writer sets the cache entry while once the cache entries are written unlimited number of readers can access the cache. 

This can be achieved using a `Lazy<T>` with a value factory delegate including `ExecutionAndPublication`. The value factory delegate still closure allocates, but that only happens when the lazy is materialized at init time or when the entry is expired. 

The alternative would be to take a dependency now to the original package as a package dependency. But I must say I find the term rate limiting rather confusing how it was used here. And if we really need rate limiting (which I doubt based on what I heard from @mauroservienti ) then I think we should be using [System.Threading.RateLimiting](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/) instead.

Here is some profiling data from running the one event publish tests. The effect of this change should be even more "drastic" for endpoints that publish multiple events.

## Before

Time spent in the rate-limiting code:

image

image

## After

image

### CODE REVIEW

**Comment by danielmarbach** [comment]
Now that I better understand the problem here I think we can remove the pulled in code and replace it with a cache that uses `Lazy` similar to what I have done here https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/blob/master/src/Transport/Sending/MessageSenderRegistry.cs#L24-L32

**Review by mauroservienti — APPROVED** [review]

**Comment by danielmarbach** [comment]
Interesting. Running now into some sort of exhaustion problem on Windows with net6.

**Comment by danielmarbach** [comment]
Interesting the latest commit [made everything green](https://github.com/Particular/NServiceBus.AmazonSQS/actions/runs/4275382253) (which I can't explain). Going to run one more time because before I consistently ran into the exhausting exception.

**Review by jpalac — APPROVED** [review]

File: src/NServiceBus.Transport.SQS/HybridPubSubChecker.cs [hunk 1]
```diff
@@ -71,34 +48,53 @@ public async Task<bool> ThisIsAPublishMessageNotUsingMessageDrivenPubSub(Unicast
...
                 var cacheKey = existingTopic.TopicArn + unicastTransportOperation.Destination;
-                Logger.Debug($"Performing first subscription cache lookup for '{cacheKey}'.");
-                if (!TryGetFromCache(cacheKey, out var cacheItem))
+                if (Logger.IsDebugEnabled)
                 {
-                    cacheItem = await rateLimiter.Execute(async () =>
+                    Logger.Debug($"Performing subscription cache lookup for '{cacheKey}'.");
+                }
+
+                var lazyCacheItem = subscriptionsCache.AddOrUpdate(cacheKey,
```
Comment (tmasternak) [inline src/NServiceBus.Transport.SQS/HybridPubSubChecker.cs:63]: Do I understand it correctly, that `AddOrUpdate` will synchronize the creation of the `LazyCacheItem`s. Eg. if there are two queries for the same topic, the first to execute `AddOrUpdate` will schedule `LazyCacheItem` creation, and the second one will get to the `Update` lambda, check that the entry is fresh, and return what the first has grabbed from SNS.

File: src/NServiceBus.Transport.SQS.Tests/TopicCacheTests.cs [hunk 2]
```diff
@@ -30,6 +30,143 @@ public async Task GetTopicArn_caches()
...
+            snsClient.FindTopicRequests.Clear();
+
+            await Task.Delay(500);
+
+            snsClient.FindTopicAsyncResponse = original;
+
+            result1 = await cache.GetTopicArn(typeof(Event));
+            result2 = await cache.GetTopicArn(typeof(Event));
+
+            Assert.That(result1, Is.EqualTo("arn:aws:sns:us-west-2:123456789012:PREFIXEvent"));
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Transport.SQS.Tests/TopicCacheTests.cs:60]: Yes because that is built into the fake :dancers:

**Comment by danielmarbach** [comment]
Merging for now. Should additional reviews discover things to change, we can do follow-up PRs.

**Pull request merged**